File: apps/api/services/feature_flag_migration.py
from __future__ import annotations

import hashlib
import json
from typing import Any, cast

from sqlalchemy import select, text
from sqlalchemy.ext.asyncio import AsyncSession

from core.config import get_settings
from core.id import generate_id
from core.timestamps import now_unix_seconds
from db.models import (
    Feature,
    FeatureFlagMigrationArchive,
)
from providers.posthog.client import get_posthog_client


async def capture_legacy_feature_snapshot(
    db: AsyncSession,
) -> FeatureFlagMigrationArchive:
    """Persist a complete, checksummed legacy-provider and entitlement snapshot."""

    features = await _table_rows(db, "features")
    user_grants = await _table_rows(db, "user_features")
    org_grants = await _table_rows(db, "org_features")
    user_identities = await _legacy_user_identity_rows(db)

    snapshot_data = {
        "schema_version": 1,
        "legacy_provider": {"name": "removed"},
        "local": {
            "features": features,
            "user_features": user_grants,
            "org_features": org_grants,
            "user_identities": user_identities,
        },
    }
    checksum = hashlib.sha256(_canonical_json(snapshot_data).encode()).hexdigest()
    snapshot = _normalize_json(
        {**snapshot_data, "captured_at": now_unix_seconds()}
    )
    existing = (
        await db.scalars(
            select(FeatureFlagMigrationArchive).where(
                FeatureFlagMigrationArchive.checksum == checksum
            )
        )
    ).first()
    if existing:
        return existing

    counts = {
        "local_features": len(features),
        "local_user_grants": len(user_grants),
        "local_org_grants": len(org_grants),
        "local_user_identities": len(user_identities),
    }
    archive = FeatureFlagMigrationArchive(
        id=generate_id("featureFlagMigrationArchive"),
        source_provider="legacy",
        target_provider="posthog",
        checksum=checksum,
        status="captured",
        counts=counts,
        snapshot=snapshot,
        result=None,
        created_at=now_unix_seconds(),
        completed_at=None,
    )
    db.add(archive)
    await db.flush()
    return archive


async def import_snapshot_to_posthog(
    db: AsyncSession,
    archive: FeatureFlagMigrationArchive,
) -> dict[str, Any]:
    """Idempotently create PostHog flags, then switch local provider mappings."""

    settings = get_settings()
    posthog = get_posthog_client(settings)
    existing_flags = {str(row.get("key")): row for row in await posthog.list_features()}
    local_features = list(
        (await db.scalars(select(Feature).order_by(Feature.created_at, Feature.id))).all()
    )
    mappings: list[dict[str, str]] = []
    archived_features = {
        str(row.get("id")): row
        for row in archive.snapshot.get("local", {}).get("features", [])
    }

    for feature in local_features:
        provider_flag = existing_flags.get(feature.slug)
        if provider_flag is None:
            provider_flag = await posthog.create_feature(
                key=feature.slug,
                name=feature.name,
                description=feature.description,
                enabled=feature.enabled,
            )
            existing_flags[feature.slug] = provider_flag

        provider_id = str(provider_flag["id"])
        mappings.append(
            {
                "feature_id": feature.id,
                "slug": feature.slug,
                "source_provider_feature_id": str(
                    archived_features.get(feature.id, {}).get("provider_feature_id") or ""
                ),
                "posthog_feature_id": provider_id,
            }
        )

    if len(mappings) != len(local_features):
        raise RuntimeError("PostHog migration did not map every local feature.")

    for feature in local_features:
        provider_flag = existing_flags[feature.slug]
        feature.provider = "posthog"
        feature.provider_feature_id = str(provider_flag["id"])
        feature.provider_environment_id = str(settings.posthog_project_id)
        feature.provider_metadata = provider_flag
        feature.synced_at = now_unix_seconds()
        feature.updated_at = now_unix_seconds()

    result = {
        "archive_id": archive.id,
        "checksum": archive.checksum,
        "mapped_features": len(mappings),
        "mappings": mappings,
    }
    archive.status = "completed"
    archive.result = result
    archive.completed_at = now_unix_seconds()
    await db.flush()
    return result


async def remove_legacy_feature_columns(db: AsyncSession) -> None:
    """Remove obsolete provider columns only after the archive/import commits."""

    columns_by_table = {
        "features": (
            "flagsmith_feature_id",
            "flagsmith_owners",
            "flagsmith_created_at",
            "flagsmith_updated_at",
        ),
        "user_features": ("flagsmith_identity_id", "flagsmith_featurestate_id"),
        "org_features": ("flagsmith_identity_id", "flagsmith_featurestate_id"),
        "users": (
            "flagsmith_identity_id",
            "flagsmith_identifier",
            "flagsmith_environment_id",
            "flagsmith_identity_synced_at",
        ),
    }
    for table_name, columns in columns_by_table.items():
        actions = ", ".join(f"DROP COLUMN IF EXISTS {column}" for column in columns)
        await db.execute(text(f"ALTER TABLE {table_name} {actions}"))


async def _table_rows(db: AsyncSession, table_name: str) -> list[dict[str, Any]]:
    result = await db.execute(
        text(f"SELECT to_jsonb(row_data) AS data FROM {table_name} AS row_data ORDER BY id")
    )
    return [dict(row.data) for row in result if isinstance(row.data, dict)]


async def _legacy_user_identity_rows(db: AsyncSession) -> list[dict[str, Any]]:
    columns = set(
        (
            await db.execute(
                text(
                    "SELECT column_name FROM information_schema.columns "
                    "WHERE table_name = 'users'"
                )
            )
        ).scalars()
    )
    legacy_columns = [
        "flagsmith_identity_id",
        "flagsmith_identifier",
        "flagsmith_environment_id",
        "flagsmith_identity_synced_at",
    ]
    if not set(legacy_columns).issubset(columns):
        return []

    projection = ", ".join(["id", *legacy_columns])
    result = await db.execute(text(f"SELECT {projection} FROM users ORDER BY id"))
    return [dict(row._mapping) for row in result]


def _normalize_json(value: Any) -> dict[str, Any]:
    return cast(dict[str, Any], json.loads(json.dumps(value, default=str)))


def _canonical_json(value: dict[str, Any]) -> str:
    return json.dumps(value, sort_keys=True, separators=(",", ":"))

File: apps/api/src/workers/feature-flag-migration.ts
import { createHash } from 'node:crypto'

import { getSettings } from '@/config'
import { getLogger } from '@/platform/logger'
import { generateId } from '@/platform/ids'
import { nowUnixSeconds } from '@/platform/timestamps'
import { getPostHogClient } from '@/providers/posthog/client'

import {
  createArchive,
  findArchiveByChecksum,
  findCompletedArchive,
  findArchiveById,
  findLatestArchive,
  legacyUserIdentityRows,
  listFeatures,
  tableRows,
  updateArchive,
  updateFeatureProvider,
} from './feature-flag-migration.repository'

const log = getLogger('feature-flag-migration')

export type FeatureFlagMigrationSummary = {
  archiveId: string
  checksum: string
  status: string
  mappedFeatures: number
  configured: boolean
}

function canonicalJson(value: unknown): string {
  return JSON.stringify(value, (_key, val) => {
    if (val !== null && typeof val === 'object' && !Array.isArray(val)) {
      const sorted: Record<string, unknown> = {}
      for (const key of Object.keys(val as Record<string, unknown>).sort()) {
        sorted[key] = (val as Record<string, unknown>)[key]
      }
      return sorted
    }
    return val
  })
}

function _normalizeJson(value: unknown): Record<string, unknown> {
  return JSON.parse(
    JSON.stringify(value, (_key, val) => {
      if (typeof val === 'bigint') return String(val)
      if (
        val !== null &&
        typeof val === 'object' &&
        typeof (val as { toString?: () => string })?.toString === 'function'
      ) {
        // default=str in Python handles non-serializable types; for our snapshot
        // all values are JSON-serializable, so passthrough.
      }
      return val
    })
  ) as Record<string, unknown>
}

export async function captureLegacyFeatureSnapshot(): Promise<{
  archiveId: string
  checksum: string
  created: boolean
}> {
  const features = await tableRows('features')
  const userGrants = await tableRows('user_features')
  const orgGrants = await tableRows('org_features')
  const userIdentities = await legacyUserIdentityRows()

  const snapshotData = {
    schema_version: 1,
    legacy_provider: { name: 'removed' },
    local: {
      features,
      user_features: userGrants,
      org_features: orgGrants,
      user_identities: userIdentities,
    },
  }

  const canonical = canonicalJson(snapshotData)
  const checksum = createHash('sha256').update(canonical).digest('hex')

  const existing = await findArchiveByChecksum(checksum)
  if (existing) {
    return { archiveId: existing.id, checksum, created: false }
  }

  const counts = {
    local_features: features.length,
    local_user_grants: userGrants.length,
    local_org_grants: orgGrants.length,
    local_user_identities: userIdentities.length,
  }

  const snapshot = {
    ...snapshotData,
    captured_at: nowUnixSeconds(),
  }

  // Normalize via JSON round-trip with default=str — handles BigInt etc.
  const normalizedSnapshot = _normalizeJson(snapshot)

  const archiveId = generateId('featureFlagMigrationArchive')
  await createArchive({
    id: archiveId,
    sourceProvider: 'legacy',
    targetProvider: 'posthog',
    checksum,
    status: 'captured',
    counts,
    snapshot: normalizedSnapshot,
    result: null,
    createdAt: BigInt(nowUnixSeconds()),
    completedAt: null,
  })

  log.info(
    { archive_id: archiveId, checksum, counts },
    'feature_flag_migration.captured'
  )

  return { archiveId, checksum, created: true }
}

export async function importSnapshotToPosthog(
  archiveId?: string | null
): Promise<FeatureFlagMigrationSummary> {
  const settings = getSettings()
  if (
    !settings.posthog.personalApiKey ||
    !settings.posthog.projectId ||
    !settings.posthog.host
  ) {
    return {
      archiveId: archiveId ?? '',
      checksum: '',
      status: 'skipped',
      mappedFeatures: 0,
      configured: false,
    }
  }

  let archive: Awaited<ReturnType<typeof findLatestArchive>> = null
  // A named archive wins; without one, the most recent capture is replayed.
  if (archiveId) {
    archive = (await findArchiveById(archiveId)) as unknown as typeof archive
  }
  if (!archive) {
    archive = await findLatestArchive()
  }
  if (!archive) {
    throw new Error(
      'No feature flag migration archive found. Run capture first.'
    )
  }

  const posthog = getPostHogClient(settings)
  const existingFlagsRaw = await posthog.listFeatures()
  const existingFlags = new Map<string, Record<string, unknown>>()
  for (const flag of existingFlagsRaw) {
    const key = String(flag['key'] ?? '')
    if (key) existingFlags.set(key, flag)
  }

  const localFeatures = await listFeatures()
  const snapshot = archive.snapshot as Record<string, unknown>
  const snapshotLocal =
    (snapshot['local'] as Record<string, unknown> | undefined) ?? {}
  const snapshotFeatures =
    (snapshotLocal['features'] as Array<Record<string, unknown>> | undefined) ??
    []
  const archivedFeatures = new Map<string, Record<string, unknown>>()
  for (const row of snapshotFeatures) {
    const id = String(row['id'] ?? '')
    if (id) archivedFeatures.set(id, row)
  }

  const mappings: Array<{
    feature_id: string
    slug: string
    source_provider_feature_id: string
    posthog_feature_id: string
  }> = []

  for (const feature of localFeatures) {
    let providerFlag = existingFlags.get(feature.slug) ?? null
    if (!providerFlag) {
      providerFlag = await posthog.createFeature({
        key: feature.slug,
        name: feature.name,
        description: feature.description,
        enabled: feature.enabled,
      })
      existingFlags.set(feature.slug, providerFlag)
    }

    const providerId = String(providerFlag['id'])
    const archived = archivedFeatures.get(feature.id)
    const sourceProviderFeatureId = String(
      archived?.['provider_feature_id'] ?? archived?.['providerFeatureId'] ?? ''
    )

    mappings.push({
      feature_id: feature.id,
      slug: feature.slug,
      source_provider_feature_id: sourceProviderFeatureId,
      posthog_feature_id: providerId,
    })
  }

  if (mappings.length !== localFeatures.length) {
    throw new Error('PostHog migration did not map every local feature.')
  }

  const now = BigInt(nowUnixSeconds())
  for (const feature of localFeatures) {
    const providerFlag = existingFlags.get(feature.slug)
    if (!providerFlag) continue
    await updateFeatureProvider(feature.id, {
      provider: 'posthog',
      providerFeatureId: String(providerFlag['id']),
      providerEnvironmentId: String(settings.posthog.projectId),
      providerMetadata: providerFlag,
      syncedAt: now,
      updatedAt: now,
    })
  }

  const result = {
    archive_id: archive.id,
    checksum: archive.checksum,
    mapped_features: mappings.length,
    mappings,
  }

  await updateArchive(archive.id, {
    status: 'completed',
    result,
    completedAt: now,
  })

  log.info(
    { archive_id: archive.id, mapped_features: mappings.length },
    'feature_flag_migration.completed'
  )

  return {
    archiveId: archive.id,
    checksum: archive.checksum,
    status: 'completed',
    mappedFeatures: mappings.length,
    configured: true,
  }
}

export async function runFeatureFlagMigrationOnce(): Promise<FeatureFlagMigrationSummary> {
  const completed = await findCompletedArchive()
  if (completed) {
    return {
      archiveId: completed.id,
      checksum: completed.checksum,
      status: completed.status,
      mappedFeatures: 0,
      configured: true,
    }
  }

  const { archiveId } = await captureLegacyFeatureSnapshot()
  return importSnapshotToPosthog(archiveId)
}

function sleepWithAbort(ms: number, signal?: AbortSignal): Promise<void> {
  if (!signal) return new Promise((resolve) => setTimeout(resolve, ms))
  if (signal.aborted) return Promise.resolve()
  return new Promise<void>((resolve) => {
    const timeout = setTimeout(() => {
      signal.removeEventListener('abort', onAbort)
      resolve()
    }, ms)
    const onAbort = (): void => {
      clearTimeout(timeout)
      signal.removeEventListener('abort', onAbort)
      resolve()
    }
    signal.addEventListener('abort', onAbort, { once: true })
  })
}

export async function runFeatureFlagMigrationWorker(options?: {
  signal?: AbortSignal
  intervalMs?: number
}): Promise<void> {
  const intervalMs = options?.intervalMs ?? 60_000

  while (!options?.signal?.aborted) {
    try {
      await runFeatureFlagMigrationOnce()
    } catch (error) {
      log.error({ err: error }, 'feature_flag_migration.worker_failed')
    }

    if (options?.signal?.aborted) break
    await sleepWithAbort(intervalMs, options?.signal)
  }
}

export function startFeatureFlagMigrationWorker(options?: {
  signal?: AbortSignal
  intervalMs?: number
}): { stop: () => void } {
  const controller = new AbortController()
  const externalSignal = options?.signal

  if (externalSignal) {
    if (externalSignal.aborted) controller.abort()
    else {
      externalSignal.addEventListener('abort', () => controller.abort(), {
        once: true,
      })
    }
  }

  void runFeatureFlagMigrationWorker({
    signal: controller.signal,
    intervalMs: options?.intervalMs,
  })

  return {
    stop: () => controller.abort(),
  }
}
